### PR TITLE
Add RTCIceCandidateInit and allow setting sdpMid and sdpMLineIndex

### PR DESCRIPTION
**Pull request by sdb9696**
# Proposed Changes

- Add `RTCIceCandidateInit` with `sdpMid` and `sdpMLineIndex` to allow passing these values with the candidate.
- Deprecate `RTCIceCandidate` in favor of `RTCIceCandidateInit`

### CODE REVIEW

File: uv.lock [hunk 1]
```diff
@@ -383,7 +383,7 @@ wheels = [
 
 [[package]]
 name = "webrtc-models"
-version = "0.0.0"
+version = "0.3.0"
```
Comment (edenhaus) [inline uv.lock:386]: Revert this change, the version will by set by the CI on the release workflow

File: pyproject.toml [hunk 2]
```diff
@@ -21,7 +21,7 @@ dependencies = [
     "mashumaro~=3.13",
     "orjson>=3.10.7",
 ]
-version = "0.0.0"
+version = "0.3.0"
```
Comment (edenhaus) [inline pyproject.toml:24]: Revert this change, the version will by set by the CI on the release workflow

File: webrtc_models/__init__.py [hunk 3]
```diff
@@ -56,3 +56,4 @@ class RTCIceCandidate(_RTCBaseModel):
     """
 
     candidate: str
+    sdp_m_line_index: int
```
Comment (edenhaus) [inline webrtc_models/__init__.py:59]: Please add an alias so the field is serialized as specified in the spec. See ice_servers above how to do it

**Review by edenhaus — CHANGES_REQUESTED** [review]

File: webrtc_models/__init__.py (hunk 3 above)
Comment (sdb9696) [reply webrtc_models/__init__.py:59]: Ok will do. I haven't made this optional which is different to the spec. If I make it optional I'll probably need to add sdpMid because one or the other is required. What do you think?

Comment (edenhaus) [reply webrtc_models/__init__.py]: We should be spec conform

Comment (sdb9696) [reply webrtc_models/__init__.py]: Ok that's done

File: webrtc_models/__init__.py [hunk 4]
```diff
@@ -48,11 +50,158 @@ class RTCConfiguration(_RTCBaseModel):
...
+        metadata=field_options(alias="sdpMLineIndex"), default=None, kw_only=True
+    )
+    sdp_mid: str | None = field(
+        metadata=field_options(alias="sdpMid"), default=None, kw_only=True
+    )
+    user_fragment: str | None = field(
+        metadata=field_options(alias="userFragment"), default=None, kw_only=True
+    )
+
+
```
Comment (edenhaus) [inline webrtc_models/__init__.py:71]:
```suggestion
```

File: webrtc_models/__init__.py [hunk 5]
```diff
@@ -48,11 +50,158 @@ class RTCConfiguration(_RTCBaseModel):
...
+            raise TypeError(msg)
+
+        if ric_init:
+            # Attributes are readonly so set them via __setattr__
+            object.__setattr__(self, "candidate", ric_init.candidate)
+            object.__setattr__(self, "sdp_mid", ric_init.sdp_mid)
+            object.__setattr__(self, "sdp_m_line_index", ric_init.sdp_m_line_index)
+            object.__setattr__(self, "user_fragment", ric_init.user_fragment)
+        else:
+            object.__setattr__(self, "candidate", "")
```
Comment (edenhaus) [inline webrtc_models/__init__.py:207]:
```suggestion
    candidate: str

    def __post_init__(self) -> None:
        """Initialize class."""
        msg = ("Using RTCIceCandidate is deprecated. Use "
               "RTCIceCandidateInit instead")
        warn(msg, DeprecationWarning, stacklevel=2)

@dataclass(frozen=True)
class RTCIceCandidateInit(RTCIceCandidate):
    """RTC Ice Candidate Init.

    See https://www.w3.org/TR/webrtc/#dom-rtcicecandidateinit
    """

    candidate: str
    sdp_mid: str | None = field(
        metadata=field_options(alias="sdpMid"), default=None, kw_only=True
    )
    sdp_m_line_index: int | None = field(
        metadata=field_options(alias="sdpMLineIndex"), default=None, kw_only=True
    )
    user_fragment: str | None = field(
        metadata=field_options(alias="userFragment"), default=None, kw_only=True
    )

    def __post_init__(self) -> None:
        """Initialize class.

        Spec compliance: If both the sdpMid and sdpMLineIndex members of
        candidateInitDict are null, throw a TypeError.
        """

        if self.sdp_mid is None and self.sdp_m_line_index is None:
            raise TypeError("sdpMid or sdpMLineIndex must be set")

```

Please keep it simple as we deprecate and remove the class anyway.

As `RTCIceCandidateInit` is a subclass `RTCIceCandidate` we can avoid the breaking change. See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides

Changes in core are also "minimal"

```diff
diff --git a/homeassistant/components/camera/__init__.py b/homeassistant/components/camera/__init__.py
index d31d21d424c..121b954ecee 100644
--- a/homeassistant/components/camera/__init__.py
+++ b/homeassistant/components/camera/__init__.py
@@ -20,7 +20,7 @@ from aiohttp import hdrs, web
 import attr
 from propcache import cached_property, under_cached_property
 import voluptuous as vol
-from webrtc_models import RTCIceCandidate, RTCIceServer
+from webrtc_models import RTCIceCandidateInit, RTCIceServer
 
 from homeassistant.components import websocket_api
 from homeassistant.components.http import KEY_AUTHENTICATED, HomeAssistantView
@@ -865,7 +865,7 @@ class Camera(Entity, cached_properties=CACHED_PROPERTIES_WITH_ATTR_):
         return config
 
     async def async_on_webrtc_candidate(
-        self, session_id: str, candidate: RTCIceCandidate
+        self, session_id: str, candidate: RTCIceCandidateInit
     ) -> None:
         """Handle a WebRTC candidate."""
         if self._webrtc_provider:
diff --git a/homeassistant/components/camera/webrtc.py b/homeassistant/components/camera/webrtc.py
index 0612c96e40c..a74b2dfe5ac 100644
--- a/homeassistant/components/camera/webrtc.py
+++ b/homeassistant/components/camera/webrtc.py
@@ -11,7 +11,12 @@ import logging
 from typing import TYPE_CHECKING, Any, Protocol
 
 import voluptuous as vol
-from webrtc_models import RTCConfiguration, RTCIceCandidate, RTCIceServer
+from webrtc_models import (
+    RTCConfiguration,
+    RTCIceCandidate,
+    RTCIceCandidateInit,
+    RTCIceServer,
+)
 
 from homeassistant.components import websocket_api
 from homeassistant.core import HomeAssistant, callback
@@ -78,7 +83,7 @@ class WebRTCAnswer(WebRTCMessage):
 class WebRTCCandidate(WebRTCMessage):
     """WebRTC candidate."""
 
-    candidate: RTCIceCandidate
+    candidate: RTCIceCandidate | RTCIceCandidateInit
 
     def as_dict(self) -> dict[str, Any]:
         """Return a dict representation of the message."""
@@ -146,7 +151,7 @@ class CameraWebRTCProvider(ABC):
 
     @abstractmethod
     async def async_on_webrtc_candidate(
-        self, session_id: str, candidate: RTCIceCandidate
+        self, session_id: str, candidate: RTCIceCandidateInit
     ) -> None:
         """Handle the WebRTC candidate."""
 
@@ -329,7 +334,7 @@ async def ws_candidate(
         return
 
     await camera.async_on_webrtc_candidate(
-        msg["session_id"], RTCIceCandidate(msg["candidate"])
+        msg["session_id"], RTCIceCandidateInit(msg["candidate"], sdp_mid="0")
     )
     connection.send_message(websocket_api.result_message(msg["id"]))
 
diff --git a/tests/components/camera/test_webrtc.py b/tests/components/camera/test_webrtc.py
index ba5cf35c52f..aa3e6f3ada6 100644
--- a/tests/components/camera/test_webrtc.py
+++ b/tests/components/camera/test_webrtc.py
@@ -6,7 +6,7 @@ from typing import Any
 from unittest.mock import AsyncMock, Mock, patch
 
 import pytest
-from webrtc_models import RTCIceCandidate, RTCIceServer
+from webrtc_models import RTCIceCandidate, RTCIceCandidateInit, RTCIceServer
 
 from homeassistant.components.camera import (
     DATA_ICE_SERVERS,
@@ -972,7 +972,7 @@ async def test_ws_webrtc_candidate(
         assert response["type"] == TYPE_RESULT
         assert response["success"]
         mock_on_webrtc_candidate.assert_called_once_with(
-            session_id, RTCIceCandidate(candidate)
+            session_id, RTCIceCandidateInit(candidate, sdp_mid="0")
         )
 
 
@@ -1024,7 +1024,7 @@ async def test_ws_webrtc_candidate_webrtc_provider(
         assert response["type"] == TYPE_RESULT
         assert response["success"]
         mock_on_webrtc_candidate.assert_called_once_with(
-            session_id, RTCIceCandidate(candidate)
+            session_id, RTCIceCandidateInit(candidate, sdp_mid="0")
         )
 
 

 ```

File: tests/fixtures/RTCIceCandidate_end.json [hunk 6]
```diff
@@ -1,3 +1,6 @@
 {
-  "candidate": ""
+  "candidate": "",
```
Comment (edenhaus) [inline tests/fixtures/RTCIceCandidate_end.json:2]: Please revert changes to this file and it must be valid to keep backwards compability

File: tests/fixtures/RTCIceCandidate_candidate.json [hunk 7]
```diff
@@ -1,3 +1,6 @@
 {
-  "candidate": "3932168448 1 udp 1694498815 1.2.3.4 10676 typ srflx raddr 0.0.0.0 rport 37566"
+  "candidate": "3932168448 1 udp 1694498815 1.2.3.4 10676 typ srflx raddr 0.0.0.0 rport 37566",
+  "sdpMLineIndex": 0,
```
Comment (edenhaus) [inline tests/fixtures/RTCIceCandidate_candidate.json:3]: Please revert changes to this file and it must be valid to keep backwards compability

File: tests/test_init.py [hunk 8]
```diff
@@ -44,10 +44,39 @@ def test_decoding_and_encoding(
 
     file_content_dict = orjson.loads(file_content)
     instance_dict = instance.to_dict()
+    assert instance_dict == snapshot(name="dict")
```
Comment (edenhaus) [inline tests/test_init.py:47]: Why did you add this?

**Review by edenhaus — CHANGES_REQUESTED** [review]

File: tests/fixtures/RTCIceCandidate_end.json (hunk 6 above)
Comment (sdb9696) [reply tests/fixtures/RTCIceCandidate_end.json]: Done

File: tests/fixtures/RTCIceCandidate_candidate.json (hunk 7 above)
Comment (sdb9696) [reply tests/fixtures/RTCIceCandidate_candidate.json]: Done

File: webrtc_models/__init__.py [hunk 9]
```diff
@@ -56,3 +57,39 @@ class RTCIceCandidate(_RTCBaseModel):
...
+        metadata=field_options(alias="sdpMLineIndex"), default=None, kw_only=True
+    )
+    user_fragment: str | None = field(
+        metadata=field_options(alias="userFragment"), default=None, kw_only=True
+    )
+
+    def __post_init__(self) -> None:
+        """Initialize class."""
+        if not self.candidate:
+            return
```
Comment (edenhaus) [inline webrtc_models/__init__.py:89]:
```suggestion
            # An empty string represents an end-of-candidates indication or a peer reflexive remote candidate 
            return
```

**Review by edenhaus — CHANGES_REQUESTED** [review]
Please also update the PR description

**Review by edenhaus — APPROVED** [review]
Thanks @sdb9696 :+1:
